### DISCUSSION TITLE
Schema-Only-Property Types

### DISCUSSION THREAD

**Discussion by gramian** [ArcadeData/arcadedb#1269] [Ideas]
Dear All,

It seems there ArcadeDB has currently no way to set up a type in way such that only properties can be added that were defined in a schema. And in case a record is inserted that contains a property that is not in the schema, an error is thrown; for example:
```
CREATE DOCUMENT TYPE doc;
CREATE PROPERTY doc.num LONG;
INSERT INTO doc SET str = 'hi there';
```
This works, as it is expected for the document model. I would like to propose a variant of a type definition making the `INSERT` above illegal. 
A way to do that practically would be a new attribute named ie `LOCKED`, so the following could be used:
```
ALTER TYPE doc LOCKED true;
```
After this type alteration, the above `INSERT` would throw an error.

There maybe syntactically other (better) ways to "lock" a type as well as other (better) names than `LOCKED`, please feel free to suggest alternatives.

Thank you for considering my proposal

**Comment by lvca** [comment]
This is quite simple to implement: in the validation method. I see there could be 3 different cases:
- `schema-less`, like now
- `schema-strict` like with the example you were mentioning
- `schema-hybrid` where all the defined properties must be filled (unless some properties have `default`), but additional properties are still allowed

WDYT?
